File: Day 025/Problem 1669. Merge In Between Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeInBetween(ListNode* list1, int a, int b, ListNode* list2) {
        ListNode* ptr = list1, *list1Prev = nullptr, *list1Next = nullptr;

        int count = 0;
        while (count < a - 1 && ptr) {
            ptr = ptr->next;
            count++;
        }
        list1Prev = ptr;

        while (count < b && ptr) {
            ptr = ptr->next;
            count++;
        }
        if (ptr) list1Next = ptr->next;

        // Find the last node of list2
        ListNode* list2Last = nullptr;
        ptr = list2;

        while (ptr && ptr->next) {
            ptr = ptr->next;
        }

        list2Last = ptr;

        list1Prev->next = list2;
        list2Last->next = list1Next;

        return list1;
    }
};

File: Day 089/Problem 1669. Merge In Between Linked Lists.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) {this.val = val; this.next = next;}
}

class Solution {
    public ListNode mergeInBetween(ListNode list1, int a, int b, ListNode list2) {
        ListNode sentinel = new ListNode();
        ListNode prevNode = sentinel, ptr = list1;
        int index = 0;

        while (index < a) {
            prevNode.next = ptr;
            prevNode = prevNode.next;
            ptr = ptr.next;
            index++;
        }

        while (index <= b) {
            ptr = ptr.next;
            index++;
        }

        prevNode.next = list2;
        while (prevNode.next != null) {
            prevNode = prevNode.next;
        }

        prevNode.next = ptr;
        return sentinel.next;
    }
}